### ISSUE TITLE
Order for generating the filtered files

### ISSUE THREAD

**Issue by hyanwong** [OneZoom/tree-build#15]
I wonder if, in the instructions, we should be giving the instructions for creating the filtered files immediately after downloading them (and if, for example, we want to give instructions for producing the filtered Wikidata JSON file independently of filtering the other files.

The reason I suggest this is that some of the files are so big that people might want to save the filtered versions and delete the original ASAP. For people short on disk space, I guess it might even be possible to pipe the download thought the filtering utility immediately, so that the huge intermediate file doesn't need to be stored?

**Comment by hyanwong** [comment]
Presumably something like:

```
wget -qO- https://dumps.wikimedia.org/wikidatawiki/entities/latest-all.json.bz2 | filter_wikidata_JSON.py -format> latest-all_filtered.json
```

Where `filter_wikidata_JSON.py` is a helper utility that wraps the [generate_filtered_wikidata_dump](https://github.com/OneZoom/tree-build/blob/1894048a711ded58b37d0b74303d0246ad47fcac/oz_tree_build/utilities/generate_filtered_files.py#L190) function using e.g. https://docs.python.org/3.8/library/bz2.html#bz2.BZ2Decompressor so that it can decompress on the fly

**Comment by hyanwong** [comment]
The line-by-line BZ2 decompressor in the answer at https://stackoverflow.com/questions/47778579/how-to-read-lines-from-arbitrary-bz2-streams-for-csv might be useful for this.

**Comment by davidebbo** [comment]
Yes, adding the option to process the Wikidata dump straight from the download steam is a good idea, and we should look into that.

Some things to think about:
- It removes the possibility of ever filtering it in parallel (since it's non seekable), but I don't think that was really feasible anyway.
- As I recall, the download itself is much slower than the filtering (like 24 hours vs 5 hours?). So if something goes wrong during filtering, you lose all your download work. Also, if the **download** fails part way (e.g. connection problem), you lose the opportunity to resume the download where you left off (which I think some tools like wget can do)
- It makes invalidation trickier as we can't compare the timestamp (or hash) of the original file (since we never actually store the original file). I guess that just means this would be an operation that always does the processing when you invoke it, without trying to have caching behavior.

I don't think any of these are reasons not to do this, but still worth thinking about.

**Comment by hyanwong** [comment]
Great points. 

> filtering it in parallel

completely agree

>  if something goes wrong during filtering, you lose all your download work ... opportunity to resume

I just did the JSON download and it took 5 hours. But it seems like a good use of time to use CPU for filtering, simultaneously with bandwidth limitations for download. Re: things going wrong, I think we say this is mainly an option for those with limited disk space. Re: resuming, it would theoretically be possible to somehow store the number of bytes downloaded (e.g. into a temporary file which is deleted at the end of the download): this would [allow resuming](https://stackoverflow.com/questions/3428102/how-to-resume-an-interrupted-download-part-2) (even [using wget](https://superuser.com/questions/900562/wget-starting-from-arbitrary-byte)), but I think that's too much complication for the moment.

> makes invalidation trickier

Yep, although we can calculate the md5 hash as we stream: https://stackoverflow.com/questions/41350054/calculate-md5-on-the-fly-while-reading-large-text-file

If this isn't too much work (and I suspect it's fine), I think it might indeed be useful to do this. We could print out the hash at the end, or even save the md5 hash in the name of the filtered file, I suppose? It would still be difficult to work out invalidation, but having the hash means you can check the download worked, etc.

**Comment by lentinj** [comment]
(incoming sticky beak)

If you're supporting a pipe from stdin (which sounds like a good plan if you can regardless), instead of trying to bzip within your code why not suggest something like the following in the examples:

```
wget -qO- https://dumps.wikimedia.org/wikidatawiki/entities/latest-all.json.bz2 | bzip2 -dc | filter_wikidata_JSON.py -
```

Then there isn't an extra code-path to worry about.

Repeating downloads just because the script failed seems a somewhat rude usage of wikimedia's bandwidth, but OTOH I could be just showing my age, and in reality they don't care because cloudflare/summat makes it all happen for free. It's not as if this is a sensible thing to do during development.

> It makes invalidation trickier as we can't compare the timestamp (or hash) of the original file

One option is the ``ETag`` header, which gives you a hash of the file before you've even started:

```
$ curl -s --head https://dumps.wikimedia.org/wikidatawiki/entities/latest-all.json.bz2 | grep -i etag
ETag: "6477cc14-132548babf"
```

...but you'd have to do this as a separate operation to the pipe above.

**Comment by hyanwong** [comment]
Neat, thanks for the sticky beak @lentinj . Really useful.

**Comment by davidebbo** [comment]
Thanks @lentinj, that is definitely doable. Everything in the current code is already doing line by line processing, so changing from a file to stdin should not be hard. Though it does mean reworking the flow, since right now the tool does the filtering for multiple files in a pipeline, and we only have one stdin!

And using the ETag is a also good idea. As you point out, the filtering logic itself cannot get the ETag from the stdin pipe, so it'd have to be separate (and is more of a nice-to-have anyway).

Maybe an alternative to stdin is to support URLs on the command line as an alternative to local file names. We can have logic that takes that string and returns a stream, abstracting out all differences between a local file and a downloaded file. That would let us theoretically handle multiple downloaded files in the same command.

**Comment by lentinj** [comment]
Yeah, they're 2 different directions you can take this, either seems a valid approach. I think I'd choose depending on how likely you are to change the filtering logic. If it's pretty set in stone then something that takes a URL as an argument and downloads-extracts-and-filters automatically does seem more friendly. Also:

* You could store metadata somewhere (initial entry, separate file, in the output filename), that includes the git version hash of the script & the E-tag of the initially downloaded file. That'd make it easy to decide if it needs re-running.
* You could keep a ".reject.json" of the lines you failed to process, and keep going regardless. Then you don't have to re-download in this case, but just re-run against this file.

**Comment by davidebbo** [comment]
Thanks @lentinj. I think we can start without worrying about the failure case initially, and revisit if needed.
